cpp/src/arrow/ipc/writer.cc

```diff
@@ -182,25 +183,32 @@ class RecordBatchSerializer {
   Status CompressBodyBuffers() {
     std::unique_ptr<util::Codec> codec;
 
-    if (!(options_.compression == Compression::LZ4_FRAME ||
-          options_.compression == Compression::ZSTD)) {
-      return Status::Invalid("Only LZ4_FRAME and ZSTD compression allowed");
-    }
+    RETURN_NOT_OK(internal::CheckCompressionSupported(options_.compression));
 
     // TODO check allowed values for compression?
     AppendCustomMetadata("ARROW:experimental_compression",
                          util::Codec::GetCodecAsString(options_.compression));
 
     ARROW_ASSIGN_OR_RAISE(
         codec, util::Codec::Create(options_.compression, options_.compression_level));
-    // TODO: Parallelize buffer compression
-    for (size_t i = 0; i < out_->body_buffers.size(); ++i) {
+
+    auto CompressOne = [&](size_t i) {
       if (out_->body_buffers[i]->size() > 0) {
         RETURN_NOT_OK(
             CompressBuffer(*out_->body_buffers[i], codec.get(), &out_->body_buffers[i]));
       }
+      return Status::OK();
+    };
+
+    if (options_.use_threads) {
+      return ::arrow::internal::ParallelFor(static_cast<int>(out_->body_buffers.size()),
+                                            CompressOne);
+    } else {
+      for (size_t i = 0; i < out_->body_buffers.size(); ++i) {
+        RETURN_NOT_OK(CompressOne(i));
+      }
+      return Status::OK();
```

Comment:
This "optional parallelism" pattern occurs frequently, I'll open a JIRA about factoring it out into a helper function. 

https://issues.apache.org/jira/browse/ARROW-8299